Archive/Imo/Imo1975Q3.lean

```diff
@@ -0,0 +1,195 @@
+
+/-- `shrink θ₀ θ₁ := sin θ₀ / sin θ₁`. -/
+@[simp] def shrink (θ₀ θ₁ : ℝ) : ℝ := Real.sin θ₀ / Real.sin θ₁
+
+noncomputable section
+variable (z : ℂ)
+
+/-- Base points. -/
+@[simp] def A : Point := 0
+@[simp] def B : Point := 1
+@[simp] def C : Point := z
+
+/-- `P`: from `B` towards `C`,
+rotate by `−45°`, scale by `shrink(30°,105°)`, then translate to `B`. -/
+noncomputable def P : Point :=
+  (z - B) * Complex.exp (deg2rad (-45) * Complex.I) * shrink (deg2rad 30) (deg2rad 105) + B
+
+/-- `Q`: from `A` towards `C`,
+rotate by `+45°`, scale by `shrink(30°,105°)`, then translate to `A`. -/
+noncomputable def Q : Point :=
+  (z - A) * Complex.exp (deg2rad 45 * Complex.I) * shrink (deg2rad 30) (deg2rad 105) + A
+
+/-- `R`: from `A` towards `B`,
+rotate by `−15°`, scale by `shrink(15°,150°)`, then translate to `A`. -/
+noncomputable def R : Point :=
+  (B - A) * Complex.exp (deg2rad (-15) * Complex.I) * shrink (deg2rad 15) (deg2rad 150) + A
+
+
+/-! ### Trigonometric helper lemmas -/
+@[grind]
+lemma sqrt_two_times_sqrt_three : √2 * √3 = √6 := by
+  rw [← sqrt_mul (Nat.ofNat_nonneg _)]
+  ring_nf
+/-- `cos(π/12) = (√6 + √2)/4`. -/
+@[simp]
+lemma cos_pi_div_twelve : Real.cos (π / 12) = (√6 + √2) / 4 := by
+  have pi_div_twelve : π / 12 = π / 4 - π / 6 := by
+    ring_nf
+  simp [pi_div_twelve, Real.cos_sub]
+  grind
+/-- `sin(π/12) = (√6 − √2)/4`. -/
+@[simp]
+lemma sin_pi_div_twelve : Real.sin (π / 12) = (√6 - √2) / 4 := by
+  have pi_div_twelve : π / 12 = π / 3 - π / 4 := by
+    ring_nf
+  simp [pi_div_twelve, Real.sin_sub]
+  grind
+/-- `sin(7π/12) = (√6 + √2)/4`. -/
+@[simp]
+lemma sin_seven_pi_div_twelve : Real.sin (7 * π / 12) = (√6 + √2) / 4 := by
+  have seven_pi_div_twelve : 7 * π / 12 = π / 3 + π / 4 := by
+    ring_nf
+  simp [seven_pi_div_twelve, Real.sin_add]
+  grind
+@[simp]
+lemma complex_sin_five_pi_div_six : sin (π * 5 / 6 : ℂ) = (1 / 2 : ℂ) := by
+  norm_cast
+  rw [← Real.sin_pi_sub]
+  ring_nf
+  field_simp
+@[simp]
+lemma complex_sin_pi_div_six : sin (π / 6 : ℂ) = 1 / 2 := by
+  norm_cast
+  simp
+lemma mul_I_into_exp_neg_quarter_pi (x : ℂ) :
+    x * I * cexp (π * I * (-1 / 4)) = x * cexp (π * I * (1 / 4)) := by
+  rw [mul_assoc, mul_eq_mul_left_iff]
+  field_simp [Complex.ext_iff, exp_re, exp_im]
+  simp [neg_div, Real.sin_neg]
+/-- Algebraic identity used to rationalize and collect coefficients:
+`2√6 + 2√2 = 8 / (√6 − √2)` (as complex numbers). -/
+lemma sqrt_combo_rationalize :
+    √6 * 2 + √2 * 2 = 8 / (√6 - √2) := by
+  have : √6 - √2 ≠ 0 := by
+    simp [sub_eq_zero]
+  field_simp
+  ring_nf
+  norm_num
+
+/-! ### Core geometric step -/
+
+/-- `(Q z) − R = ((P z) − R) · e^{iπ/2}`.
+Interpretation: the vector `QR` is a 90° CCW rotation of `PR`. -/
+lemma PQR_rot90 (z : ℂ) :
+    Q z - R = (P z - R) * Complex.exp (π / 2 * I) := by
+  -- expand definitions and normalize
+  simp [Q, P, R, A, B]
+  simp [mul_comm]
+  ring_nf
+  -- push `I` into the exponential
+  rw [mul_I_into_exp_neg_quarter_pi, mul_comm]
```

Comment:
Please avoid non-terminal `simp` (that is, `simp`, as opposed to `simp only`, followed by use of a rigid tactic such as `rw`) anywhere in the proofs.